### PR TITLE
Integrate fused moe kernel with qwen3 coder

### PR DESCRIPTION
**Pull request by vanbasten23**
…rd model.

# Description

Start with a short description of what the PR does and how this is a change from
the past.

The rest of the description includes relevant details and context, examples:
- why is this change being made,
- the problem being solved and any relevant context,
- why this is a good solution,
- some information about the specific implementation,
- shortcomings of the solution and possible future improvements.

If the change fixes a Github issue, please include a link, e.g.,:
FIXES: #123456

# Tests

Please describe how you tested this change, and include any instructions and/or
commands to reproduce.

# Checklist

Before submitting this PR, please make sure:
- I have performed a self-review of my code.
- I have necessary comments in my code, particularly in hard-to-understand areas.
- I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

File: tpu_inference/layers/vllm/quantization/fp8.py [hunk 1]
```diff
@@ -249,18 +249,7 @@ def __init__(self,
...
-                bt=256,
-                bf=1024,
-                bd1=1024,
-                bd2=1024,
-                btc=256,
-                bfc=1024,
-                bd1c=1024,
-                bd2c=1024,
-            )
+            self.extra_backend_kwargs = dict(ep_axis_name=ep_axis_name, )
```
Comment (kyuyeunk) [inline tpu_inference/layers/vllm/quantization/fp8.py:252]: can you make similar changes to all the files that calls fused moe kernel? example: https://github.com/vllm-project/tpu-inference/blob/c83ae8d2a7825cdef5aa6329544f9a4110725262/tpu_inference/layers/vllm/quantization/mxfp4.py#L121-L129

File: tpu_inference/layers/vllm/quantization/compressed_tensors/compressed_tensors_moe.py [hunk 2]
```diff
@@ -98,10 +97,6 @@ def __init__(
         self.moe_backend = select_moe_backend(self.moe)
 
         self.extra_backend_kwargs = {}
```
Comment (kyuyeunk) [inline tpu_inference/layers/vllm/quantization/compressed_tensors/compressed_tensors_moe.py:99]: i don't think you can delete this? don't you need to replace `raise NotImplementedError` to something like `self.extra_backend_kwargs = dict(ep_axis_name=ep_axis_name, )`

File: .buildkite/pipeline_jax.yml [hunk 3]
```diff
@@ -294,6 +294,22 @@ steps:
...
+     key: ${KEY_PREFIX}test_17
+     soft_fail: true
+     env:
+       IS_FOR_V7X: "${IS_FOR_V7X}"
+     agents:
+       queue: ${TPU_QUEUE_SINGLE:-tpu_v6e_8_queue}
+     commands:
+       - |
+         if [[ "$$IS_FOR_V7X" == "true" ]]; then
+           .buildkite/scripts/run_in_docker.sh bash /workspace/tpu_inference/tests/e2e/check_lm_eval.sh "Qwen/Qwen3-Coder-480B-A35B-Instruct-FP8"
```
Comment (kyuyeunk) [inline .buildkite/pipeline_jax.yml]: just realized that we don't have similar test for gpt-oss. it can be done in other pr, but can you create a similar change that check accuracy for gpt-oss as well?

File: tests/e2e/check_lm_eval.sh [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+set -e # Exit immediately if a command exits with a non-zero status.
+
+if [ -z "$1" ]; then
+    echo "Usage: $0 <model_name>. A model name is needed."
+    exit 1
+fi
+MODEL_NAME=$1
+
+model_args_json=$(printf '{"pretrained": "%s", "tensor_parallel_size": 8, "max_model_len": 2048, "max_num_batched_tokens": 2048, "max_gen_toks": 256, "enable_expert_parallel": 1}' "$MODEL_NAME")
+output=$(USE_MOE_EP_KERNEL=1 VLLM_DISABLE_SHARED_EXPERTS_STREAM=1 MODEL_IMPL_TYPE=vllm lm_eval \
```
Comment (kyuyeunk) [inline tests/e2e/check_lm_eval.sh]: `VLLM_DISABLE_SHARED_EXPERTS_STREAM=1` shouldn't be needed anymore: https://github.com/vllm-project/tpu-inference/pull/1021

File: tests/e2e/check_lm_eval.sh [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+set -e # Exit immediately if a command exits with a non-zero status.
+
+if [ -z "$1" ]; then
+    echo "Usage: $0 <model_name>. A model name is needed."
+    exit 1
+fi
+MODEL_NAME=$1
+
+model_args_json=$(printf '{"pretrained": "%s", "tensor_parallel_size": 8, "max_model_len": 2048, "max_num_batched_tokens": 2048, "max_gen_toks": 256, "enable_expert_parallel": 1}' "$MODEL_NAME")
```
Comment (kyuyeunk) [inline tests/e2e/check_lm_eval.sh]: I think it might be a good idea to make this test more generic and not hard code to specific config (like `enable_expert_parallel`, `USE_MOE_EP_KERNEL`) or specific threshold.

can you modify this code so that those configs are passed via argument instead?

File: tests/e2e/check_lm_eval.sh [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    --tasks gsm8k_cot \
+    --batch_size auto \
+    --apply_chat_template \
+    --num_fewshot 8)
+
+echo "Evaluation output:"
+echo "$output"
+
+# Thresholds
+flex_threshold=0.87
```
Comment (kyuyeunk) [inline tests/e2e/check_lm_eval.sh]: what exactly is `flex_threshold` and `strict_threshold`?

File: tpu_inference/kernels/fused_moe/v1/tuned_block_sizes.py [hunk 7]
```diff
@@ -215,4 +316,8 @@ def get_tuned_block_sizes(
         ep_size,
     )
 
+    if key not in TUNED_BLOCK_SIZES:
+        print(
```
Comment (kyuyeunk) [inline tpu_inference/kernels/fused_moe/v1/tuned_block_sizes.py]: please use `logger.warning_once` instead. example: https://github.com/vllm-project/tpu-inference/blob/c83ae8d2a7825cdef5aa6329544f9a4110725262/tpu_inference/kernels/ragged_paged_attention/v3/tuned_block_sizes.py#L4368

File: tpu_inference/layers/vllm/fused_moe.py [hunk 8]
```diff
@@ -70,6 +70,14 @@ def fused_moe_apply(
     with jax.named_scope(layer._get_name()):
         match moe_backend:
             case FusedMoEBackend.FUSED_MOE:
+                subc_quant_w1_sz = None
+                subc_quant_w2_sz = None
+                if layer.moe_quant_config.per_out_ch_quant or layer.moe_quant_config.is_block_quantized:
```
Comment (kyuyeunk) [inline tpu_inference/layers/vllm/fused_moe.py]: shouldn't it be possible to figure out a `subc_quant_w1_sz` and `subc_quant_w2_sz` by utilizing information from weight shape and scale shape? example: https://github.com/vllm-project/tpu-inference/blob/c83ae8d2a7825cdef5aa6329544f9a4110725262/tpu_inference/kernels/megablox/gmm.py#L395-L398

File: tpu_inference/layers/vllm/quantization/compressed_tensors/compressed_tensors_moe.py (hunk 2 above)
Comment (vanbasten23) [reply tpu_inference/layers/vllm/quantization/compressed_tensors/compressed_tensors_moe.py:99]: I tested this path and it doesn't necessarily need it. This is because the kernel set `ep_axis_name` with the default value "model" https://github.com/vllm-project/tpu-inference/blob/248ff5e1b2562b0100f3822b4ecb97ada76f8be5/tpu_inference/kernels/fused_moe/v1/kernel.py#L1258

But I think this is a good point and I agree with you that we'd better set it explicitly.

File: .buildkite/pipeline_jax.yml (hunk 3 above)
Comment (vanbasten23) [reply .buildkite/pipeline_jax.yml]: For sure. Will do in a follow-up pr.

~~One more thing, I just realized that the v7 machine in our CI only has 1 chip. Do you know if qwen3-coder or gpt-oss can run in v6e-8?~~

Edit: Nvm, I find v7-8 machine in the CI.

File: tests/e2e/check_lm_eval.sh (hunk 6 above)
Comment (vanbasten23) [reply tests/e2e/check_lm_eval.sh]: The lm_eval output looks like
```
|  Tasks  |Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|---------|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k_cot|      3|flexible-extract|     8|exact_match|↑  |0.8802|±  |0.0089|
|         |       |strict-match    |     8|exact_match|↑  |0.7453|±  |0.0120
```
The 'Value' shows the accuracy. 'flex_threshold` is the threshold for the 'flexible-extract' score.

File: tpu_inference/layers/vllm/fused_moe.py (hunk 8 above)
Comment (vanbasten23) [reply tpu_inference/layers/vllm/fused_moe.py]: Note, hidden_size=6144, intermediate_size=2560. Here
w13_weight: [160, 2, 6144, 2560]=[num_expert, 2, hidden_size, intermediate_size]
w2_weight: [160, 2560, 6144]=[num_expert, intermediate_size, hidden_size]
w13_weight_scale: [160, 2, 1, 1, 2560]=(num_experts, 2, hidden_size // subc_quant_w1_sz, 1, intermediate_size)
w2_weight_scale: [160, 1, 1, 6144]=(num_experts, intermediate_size // subc_quant_w2_sz, 1, hidden_size)

Can you elaborate the issue with the current implementation? In the fused moe, the spec wants to use block quant. But we decided that for now we will use per-channel quant in the fused moe kernel. Is my understanding correct?

Or do you mean something like: because w13_weight_scale.shape==[160, 2, 1, 1, 2560]==(num_experts, 2, hidden_size // subc_quant_w1_sz, 1, intermediate_size), so 1==hidden_size // subc_quant_w1_sz, so subc_quant_w1_sz should be hidden_size?

**Comment by vanbasten23** [comment]
for some reason, the lm_eval works locally but fails in the CI. Need to look into it...

[Edit]: i figured it out

File: tpu_inference/layers/vllm/fused_moe.py (hunk 8 above)
Comment (kyuyeunk) [reply tpu_inference/layers/vllm/fused_moe.py]: yes, so we can think of it like this
- `w2_weight=[160, 2560, 6144]=[num_expert, intermediate_size, hidden_size]`
- therefore, `intermediate_size=2560`
- `w2_weight_scale=[160, 1, 1, 6144]=(num_experts, intermediate_size // subc_quant_w2_sz, 1, hidden_size)`
- therefore, `intermediate_size // subc_quant_w2_sz=1`
- putting them `subc_quant_w2_sz=intermediate_size // (intermediate_size // subc_quant_w2_sz)=2560 // 1 = 2560`

also, i think the existing logic in here will break for subchannel case? because if a weight is subchannel quantized, `subc_quant_w1_sz` will not be `padded_hidden_size` and will be some different value like 512 or 256. Meanwhile,  the technique i've described above will be applicable regardless of subchannel or per-channel quantization.

reason why i prefer this approach is that i'm trying my best to utilize functional programming style where it minimizes an access to class member variable & everything can be figured out via argument's shapes, numerics, etc.

for instance, if vllm upstream changes their variable name from `layer.moe_quant_config.per_out_ch_quant` to `layer.moe_quant_config.per_out_channel_quant`, this logic will break.

But if we write logic in a way we determine `subc_quant_w2_sz` by purely shapes of weights & weight scale, it is extremely robust to upstream changes.

Comment (vanbasten23) [reply tpu_inference/layers/vllm/fused_moe.py]: Got it. Thanks for the explanation. I like the functional programming way.

File: tpu_inference/layers/vllm/quantization/mxfp4.py [hunk 9]
```diff
@@ -118,15 +118,6 @@ def __init__(
                 subc_quant_w1_sz=REQUANTIZED_BLOCK_SIZE,
```
Comment (kyuyeunk) [inline tpu_inference/layers/vllm/quantization/mxfp4.py]: you need to remove these lines to resolve ci failure.

**Review by kyuyeunk — APPROVED** [review]
